### PR TITLE
EVG-13784: monitor stale retrying jobs

### PR DESCRIPTION
**Pull request by Kimchelly**
JIRA: https://jira.mongodb.org/browse/EVG-13784

* Add a goroutine to monitor for job that are in the process of retrying but are not actively marked as retrying (i.e. the lock timeout is exceeded).
    * Add a driver method to return retryable jobs based on a filter.
* Atomically lock a retry job to a single in-memory queue's retry handler.
* Move a few random helper functions to more appropriate places.

### CODE REVIEW

File: queue/driver_mongo.go [hunk 1]
```diff
@@ -735,9 +739,106 @@ func (d *mongoDriver) Jobs(ctx context.Context) <-chan amboy.Job {
...
+			if err = iter.Decode(ji); err != nil {
+				grip.Warning(message.WrapError(err, message.Fields{
+					"id":        d.instanceID,
+					"service":   "amboy.queue.mdb",
+					"is_group":  d.opts.UseGroups,
+					"group":     d.opts.GroupName,
+					"operation": "retrying job iterator",
+					"message":   "problem reading job from cursor",
+				}))
+
```
Comment (julianedwards) [inline queue/driver_mongo.go:802]: Nit: remove blank line.

File: queue/driver_mongo.go [hunk 2]
```diff
@@ -735,9 +739,106 @@ func (d *mongoDriver) Jobs(ctx context.Context) <-chan amboy.Job {
...
+			})
+		}
+
+		grip.Error(message.WrapError(iter.Err(), message.Fields{
+			"id":        d.instanceID,
+			"service":   "amboy.queue.mdb",
+			"is_group":  d.opts.UseGroups,
+			"group":     d.opts.GroupName,
+			"operation": "retrying job iterator",
+			"message":   "database interface error",
```
Comment (julianedwards) [inline queue/driver_mongo.go:835]: Do you mean `database cursor/iterator error`?

File: queue/driver_mongo.go [hunk 3]
```diff
@@ -735,9 +739,106 @@ func (d *mongoDriver) Jobs(ctx context.Context) <-chan amboy.Job {
...
+		"retry_info.needs_retry": true,
+	}
+	d.modifyQueryForGroup(q)
+	return q
+}
+
+func (d *mongoDriver) RetryableJobs(ctx context.Context, filter RetryableJobFilter) <-chan amboy.RetryableJob {
+	jobs := make(chan amboy.RetryableJob)
+
+	go func() {
```
Comment (julianedwards) [inline queue/driver_mongo.go:759]: Should this handle panics?

File: queue/driver_test.go [hunk 4]
```diff
@@ -473,6 +473,191 @@ func (s *DriverSuite) TestJobsMethodReturnsAllJobs() {
 	s.Equal(counter, len(mocks))
 }
 
+func (s *DriverSuite) TestRetryableJobsReturnsAllRetryableJobs() {
```
Comment (julianedwards) [inline queue/driver_test.go:476]: Do we need to test with groups?

**Review by julianedwards — CHANGES_REQUESTED** [review]

File: queue/driver_mongo.go (hunk 3 above)
Comment (Kimchelly) [reply queue/driver_mongo.go:770]: I mostly just copy-pasted `Jobs()`, but I can add panic handlers to both of these.

File: queue/driver_mongo.go (hunk 1 above)
Comment (Kimchelly) [reply queue/driver_mongo.go]: Done.

File: queue/driver_mongo.go (hunk 2 above)
Comment (Kimchelly) [reply queue/driver_mongo.go]: I copied this from Jobs() but I can change both.

File: queue/driver_test.go (hunk 4 above)
Comment (Kimchelly) [reply queue/driver_test.go:514]: I changed the driver suite to run all the tests with and without groups to ensure they work the same way.

**Review by julianedwards — APPROVED** [review]

**Pull request merged**